backend/btrixcloud/migrations/migration_0039_coll_slugs.py

```diff
@@ -16,23 +21,69 @@ class Migration(BaseMigration):
     def __init__(self, mdb, **kwargs):
         super().__init__(mdb, migration_version=MIGRATION_VERSION)
 
+    async def dedup_slug(
+        self, name: str, slug_base: str, coll_id: UUID, colls_mdb
+    ) -> None:
+        """attempt to set slug, if duplicate, append suffix until a valid slug is found
+        also update original name with same suffix"""
+        slug = slug_base
+        count = 1
+
+        while True:
+            try:
+                await colls_mdb.find_one_and_update(
+                    {"_id": coll_id},
+                    {"$set": {"slug": slug}},
+                )
+                print("Slug set: " + slug)
+                break
+            except DuplicateKeyError:
+                # pylint: disable=raise-missing-from
+                count += 1
+                slug = f"{slug_base}-{count}"
+
+        if count > 1:
+            print(f"Dedup name '{name}' -> '{name} {count}'")
+            await colls_mdb.find_one_and_update(
+                {"_id": coll_id}, {"$set": {"name": f"{name} {count}"}}
+            )
+
     async def migrate_up(self):
         """Perform migration up.
 
         Add slug to collections that don't have one yet, based on name
         """
         colls_mdb = self.mdb["collections"]
+        case_insensitive_collation = Collation(locale="en", strength=1)
+
+        await colls_mdb.drop_indexes()
+
+        # set slug to random value to ensure uniqueness
+        await colls_mdb.update_many(
+            {}, [{"$set": {"slug": {"$toString": {"$rand": {}}}}}]
+        )
+
+        await colls_mdb.create_index(
+            [("oid", pymongo.ASCENDING), ("slug", pymongo.ASCENDING)],
+            unique=True,
+            collation=case_insensitive_collation,
+        )
 
-        async for coll_raw in colls_mdb.find({"slug": None}):
+        async for coll_raw in colls_mdb.find({}):
```

Comment:
We may want to keep this filter, otherwise if we set `rerun_from_migration`, we might accidentally change collection slugs being used for public collections and break peoples' links

Reply:
(maybe we can re-add the filter after running this on our own instances that ran into problems)

Reply:
The previous step randomizes the slugs to a random value, so this wouldn't work. Presumably migration is only ever run once, unless rerun_from_migration is set, but then it should still be derived from collection names, since we don't allow modifying slug separate from name. This guarantees that the slugs and names are always unique.

Reply:
Hm, yeah that's a good point. So in the vast majority of cases the links/slugs should remain the same (unless there was a sort-of duplicate name because our index wasn't previously case-insensitive for collection names iirc). Okay, just gonna close this comment then.